create-diff-object symtab

sm00th opened the pull request:
As discussed in #836 this pull request is a merge of mine and @jpoimboe's attempts to change create-diff-object to take symtable (`eu-readelf -s` output) as an input instead of an object-file.  This will make unit-tests more flexible and will allow us to cover more cases with them.

There is something to discuss though: when I tried to get a symtab that would work with preexisting test objectfiles I couldn't as I lost the config I used. Even if I rebuild all of those now so that they all work with a single vmlinux.symtab - it would be unreasonable to expect other contributors to rebuild their objectfiles using a specific config/kernel version. Should we just always use per-test symtab(and possibly symvers) files then? As I understand in most cases they can be generated from .ORIG.o (we can even do it on the fly) and only in rare cases we'll need to use vmlinux/parent module symtab.

jpoimboe commented:
Yeah, let's use a per-test symtab/symvers.  If the symtab doesn't exist, it can be auto-generated and if the symvers doesn't exist, it can be /dev/null.

jpoimboe commented on kpatch-build/kpatch-build:
```diff
@@ -773,16 +773,25 @@ for i in $FILES; do
+		else
+			KOBJFILE_NAME="${KOBJFILE%.ko}"
+			KOBJFILE_NAME="${KOBJFILE_NAME/-/_}"
+			KOBJFILE_PATH="module/$KOBJFILE"
```
I think I wrote some of this code but now it looks buggy.  Shouldn't this be `$TEMPDIR/module/$KOBJFILE`?

sm00th replied:
Will $TEMPDIR/module/$KOBFILE work in non `-s` case?

jpoimboe commented on kpatch-build/kpatch-build:
```diff
@@ -773,16 +773,25 @@ for i in $FILES; do
+			KOBJFILE_PATH="module/$KOBJFILE"
+		fi
+
+		SYMTAB="$KOBJFILE_PATH.symtab"
```
In the vmlinux case, without -s, $KOBJFILE_PATH is going to point out to a non-writeable path.  Instead it should be written somewhere in $TEMPDIR.

jpoimboe commented on kpatch-build/kpatch-build:
```diff
@@ -773,16 +773,25 @@ for i in $FILES; do
+		fi
+
+		SYMTAB="$KOBJFILE_PATH.symtab"
+		eu-readelf -s "$KOBJFILE" > "$SYMTAB"
```
I think this should be $KOBJFILE_PATH instead of $KOBJFILE.

sm00th commented:
v2:
  - switched to per-test symtab/symvers
  - added now-required elfutils to travis
  - KOBJFILE_NAME passed through basename to strip path
  - KOBJFILE_PATH explicitly in $TEMPDIR
  - SYMTAB explicitly in $TEMPDIR

jpoimboe commented on kpatch-build/kpatch-build:
```diff
@@ -773,16 +773,25 @@ for i in $FILES; do
+			KOBJFILE_PATH="${TEMPDIR}/module/$KOBJFILE"
+		fi
+
+		SYMTAB="${TEMPDIR}/${KOBJFILE_NAME}.symtab"
```
This could overwrite another symtab if there are two modules being patched which have the same name (but are in different directories).  Wouldn't break anything, but could make debugging confusing.

Maybe set SYMTAB conditionally.  In the vmlinux case it could be `$TEMPDIR/vmlinux.symtab`, and in the module case it could be `$KOBJFILE.symtab`.

sm00th commented:
v3:
 - use different symtab paths for vmlinux/module cases: `${TEMPDIR}/${KOBJFILE_NAME}.symtab` in vmlinux case; `${KOBJFILE_PATH}.symtab` in module case where `${KOBJFILE_PATH}` is explicitly in `${TEMPDIR}`
